relay-system/src/service.rs

```diff
@@ -1031,6 +1033,45 @@ pub trait Service: Sized {
     }
 }
 
+/// Keeps track of running services.
+///
+/// Exposes information about crashed services.
+#[derive(Debug, Default)]
+pub struct ServiceRunner(FuturesUnordered<JoinHandle<()>>);
+
+impl ServiceRunner {
+    /// Creates a new service runner.
+    pub fn new() -> Self {
+        Self(FuturesUnordered::new())
+    }
+
+    /// Starts a service and starts tracking its join handle, exposing an [Addr] for message passing.
+    pub fn start<S: Service>(&mut self, service: S) -> Addr<S::Interface> {
+        let (addr, rx) = channel(S::name());
+        self.spawn(service, rx);
+        addr
+    }
+
+    /// Starts a service and starts tracking its join handle, given a predefined receiver.
+    pub fn spawn<S: Service>(&mut self, service: S, rx: Receiver<S::Interface>) {
+        self.0.push(tokio::spawn(service.run(rx)));
+    }
+
+    /// Awaits until all services have finished.
+    ///
+    /// Panics if one of the spawned services has panicked.
+    pub async fn join(&mut self) {
+        while let Some(res) = self.0.next().await {
+            if let Err(e) = res {
+                if e.is_panic() {
+                    // Re-trigger panic to terminate the process:
+                    std::panic::resume_unwind(e.into_panic());
```

Comment:
I confirmed that this terminates the process by inserting an explicit `panic!()` into one of the service bodies.

Reply:
Mid-/Long-Term it would be nice if a service can specify it's desired behaviour, abort on panic/restart on panic.